File: backend/app/schemas/farmer.py
"""
Pydantic schemas for Farmer Profile, Farm, Field, Crop Cultivation, and Farm Command Center.
"""

from datetime import date, datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field as PydanticField, ConfigDict


class SoilValueWithProvenance(BaseModel):
    value: Optional[Any] = None
    provenance: str = "UNKNOWN"  # "MEASURED", "ESTIMATED", "UNKNOWN"


class SoilDataInput(BaseModel):
    ph: Optional[float] = None
    ph_provenance: str = "UNKNOWN"
    nitrogen: Optional[float] = None
    nitrogen_provenance: str = "UNKNOWN"
    phosphorus: Optional[float] = None
    phosphorus_provenance: str = "UNKNOWN"
    potassium: Optional[float] = None
    potassium_provenance: str = "UNKNOWN"
    organic_carbon: Optional[float] = None
    organic_carbon_provenance: str = "UNKNOWN"
    ec: Optional[float] = None
    ec_provenance: str = "UNKNOWN"
    texture: Optional[str] = None
    texture_provenance: str = "UNKNOWN"
    moisture: Optional[float] = None
    moisture_provenance: str = "UNKNOWN"


# --- CROP PLANTING ---

class CropPlantingBase(BaseModel):
    crop_name: str
    crop_id: Optional[str] = None
    scientific_name: Optional[str] = None
    variety: Optional[str] = None
    category: Optional[str] = None
    sowing_date: Optional[date] = None
    expected_harvest_date: Optional[date] = None
    growth_stage: Optional[str] = None
    growth_stage_source: str = "farmer"  # "farmer" or "calculated"
    cultivation_type: Optional[str] = None
    irrigation_method: Optional[str] = None
    water_availability: Optional[str] = None
    status: str = "ACTIVE"  # PLANNED, ACTIVE, HARVESTED, COMPLETED
    notes: Optional[str] = None


class CropPlantingCreate(CropPlantingBase):
    field_id: int


class CropPlantingUpdate(BaseModel):
    crop_name: Optional[str] = None
    crop_id: Optional[str] = None
    scientific_name: Optional[str] = None
    variety: Optional[str] = None
    category: Optional[str] = None
    sowing_date: Optional[date] = None
    expected_harvest_date: Optional[date] = None
    growth_stage: Optional[str] = None
    growth_stage_source: Optional[str] = None
    cultivation_type: Optional[str] = None
    irrigation_method: Optional[str] = None
    water_availability: Optional[str] = None
    status: Optional[str] = None
    notes: Optional[str] = None


class CropPlantingResponse(CropPlantingBase):
    id: int
    field_id: int
    created_at: str
    updated_at: str
    model_config = ConfigDict(from_attributes=True)


# --- FIELD ---

class FieldBase(BaseModel):
    field_name: str
    area_value: float
    area_unit: str = "acre"
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    soil_type: Optional[str] = None
    soil_test_available: bool = False
    notes: Optional[str] = None

    # Soil Data & Provenance
    ph: Optional[float] = None
    ph_provenance: str = "UNKNOWN"
    nitrogen: Optional[float] = None
    nitrogen_provenance: str = "UNKNOWN"
    phosphorus: Optional[float] = None
    phosphorus_provenance: str = "UNKNOWN"
    potassium: Optional[float] = None
    potassium_provenance: str = "UNKNOWN"
    organic_carbon: Optional[float] = None
    organic_carbon_provenance: str = "UNKNOWN"
    ec: Optional[float] = None
    ec_provenance: str = "UNKNOWN"
    texture: Optional[str] = None
    texture_provenance: str = "UNKNOWN"
    moisture: Optional[float] = None
    moisture_provenance: str = "UNKNOWN"


class FieldCreate(FieldBase):
    farm_id: int


class FieldUpdate(BaseModel):
    field_name: Optional[str] = None
    area_value: Optional[float] = None
    area_unit: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    soil_type: Optional[str] = None
    soil_test_available: Optional[bool] = None
    notes: Optional[str] = None

    ph: Optional[float] = None
    ph_provenance: Optional[str] = None
    nitrogen: Optional[float] = None
    nitrogen_provenance: Optional[str] = None
    phosphorus: Optional[float] = None
    phosphorus_provenance: Optional[str] = None
    potassium: Optional[float] = None
    potassium_provenance: Optional[str] = None
    organic_carbon: Optional[float] = None
    organic_carbon_provenance: Optional[str] = None
    ec: Optional[float] = None
    ec_provenance: Optional[str] = None
    texture: Optional[str] = None
    texture_provenance: Optional[str] = None
    moisture: Optional[float] = None
    moisture_provenance: Optional[str] = None


class FieldResponse(BaseModel):
    id: int
    farm_id: int
    field_name: str
    area_value: float
    area_unit: str
    total_area_m2: float
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    soil_type: Optional[str] = None
    soil_test_available: bool
    soil_data: Dict[str, SoilValueWithProvenance]
    notes: Optional[str] = None
    plantings: List[CropPlantingResponse] = []
    created_at: str
    updated_at: str


# --- FARM ---

class FarmBase(BaseModel):
    farm_name: str
    location_name: Optional[str] = None
    latitude: float
    longitude: float
    area_value: float
    area_unit: str = "acre"
    soil_type_manual: Optional[str] = None
    water_source: Optional[str] = None
    irrigation_method: Optional[str] = None
    ownership_type: Optional[str] = None
    notes: Optional[str] = None


class FarmCreate(FarmBase):
    pass


class FarmUpdate(BaseModel):
    farm_name: Optional[str] = None
    location_name: Optional[str] = None
    latitude: Optional[float] = None
    longitude: Optional[float] = None
    area_value: Optional[float] = None
    area_unit: Optional[str] = None
    soil_type_manual: Optional[str] = None
    water_source: Optional[str] = None
    irrigation_method: Optional[str] = None
    ownership_type: Optional[str] = None
    notes: Optional[str] = None


class FarmResponse(FarmBase):
    id: int
    farmer_id: int
    total_area_m2: float
    fields: List[FieldResponse] = []
    created_at: str
    updated_at: str


# --- FARMER PROFILE ---

class FarmerProfileBase(BaseModel):
    full_name: str
    phone: Optional[str] = None
    email: Optional[str] = None
    preferred_language: str = "en"


class FarmerProfileCreate(FarmerProfileBase):
    user_id: Optional[str] = "default_farmer"


class FarmerProfileUpdate(BaseModel):
    full_name: Optional[str] = None
    phone: Optional[str] = None
    email: Optional[str] = None
    preferred_language: Optional[str] = None


class FarmerProfileResponse(FarmerProfileBase):
    id: int
    user_id: str
    farms: List[FarmResponse] = []
    created_at: str
    updated_at: str


# --- COMMAND CENTER & AGGREGATED DASHBOARD ---

class WeatherImpactItem(BaseModel):
    crop_name: str
    field_name: str
    temperature: Optional[float] = None
    rain_forecast_mm: Optional[float] = None
    impact: str
    action: str
    status: str  # "Favorable", "Monitor", "Warning"


class SoilImpactItem(BaseModel):
    crop_name: str
    field_name: str
    ph_status: str
    texture_status: str
    n_status: str
    p_status: str
    k_status: str
    impact_text: str


class IrrigationContextItem(BaseModel):
    crop_name: str
    field_name: str
    current_moisture: Optional[float] = None
    expected_moisture: Optional[float] = None
    rainfall_forecast_mm: Optional[float] = None
    irrigation_method: Optional[str] = None
    water_availability: Optional[str] = None
    status: str  # "LOW", "NORMAL", "HIGH", "MONITOR"
    next_window: Optional[str] = None
    model_note: str  # Label e.g., "Planning estimate" or "Gallipoli ML dataset baseline"


class FertilizerContextItem(BaseModel):
    crop_name: str
    field_name: str
    n_status: str
    p_status: str
    k_status: str
    model_available: bool
    model_scope_note: str  # e.g., "Western Maharashtra regional model scope limitation"
    recommendation: Optional[str] = None
    reason: str


class PestRiskItem(BaseModel):
    crop_name: str
    field_name: str
    risk_level: str  # "Low", "Moderate", "High"
    weather_drivers: List[str] = []
    crop_stage: Optional[str] = None
    action: str
    model_scope_note: str = "Environmental weather risk indicator"


class MarketWatchItem(BaseModel):
    crop_name: str
    commodity: str
    current_price: Optional[float] = None
    trend: str  # "Increasing", "Decreasing", "Stable", "Unknown"
    change_30d_pct: Optional[float] = None
    period: str = "Last 30 days"
    market_location: Optional[str] = None


class CropTimelineItem(BaseModel):
    date_label: str
    crop_name: str
    field_name: str
    event_title: str
    reason: str
    priority: str  # "Critical", "High", "Moderate", "Info"
    source: str
    evidence_status: str


class FarmAlertItem(BaseModel):
    id: str
    priority: str  # "Critical", "High", "Moderate", "Info"
    category: str  # "Weather", "Irrigation", "Pest", "Soil", "Market", "Crop"
    title: str
    description: str
    crop_name: Optional[str] = None
    field_name: Optional[str] = None
    timestamp: str
    actionable: bool = True


class ImpactMatrixRow(BaseModel):
    crop_name: str
    field_name: str
    area_display: str
    weather: str
    soil: str
    water: str
    pest: str
    market: str
    attention_level: str  # "High", "Medium", "Low"


class DataQualitySummary(BaseModel):
    location_confidence: str
    weather_freshness: str
    soil_availability: str
    npk_availability: str
    market_status: str
    profile_completeness_pct: float
    missing_fields: List[str] = []


class TodayFarmStatus(BaseModel):
    weather_summary: str
    soil_summary: str
    water_summary: str
    market_summary: str
    action_items_count: int


class FarmDashboardResponse(BaseModel):
    farmer: FarmerProfileResponse
    location: Dict[str, Any]
    total_farm_area: float
    total_farm_area_unit: str
    active_crops_count: int
    fields_count: int
    today_status: TodayFarmStatus
    active_crop_cards: List[Dict[str, Any]]
    weather_impacts: List[WeatherImpactItem]
    soil_impacts: List[SoilImpactItem]
    irrigation_items: List[IrrigationContextItem]
    fertilizer_items: List[FertilizerContextItem]
    pest_items: List[PestRiskItem]
    market_watch: List[MarketWatchItem]
    crop_calendar_events: List[Dict[str, Any]]
    alerts: List[FarmAlertItem]
    timeline: List[CropTimelineItem]
    impact_matrix: List[ImpactMatrixRow]
    data_quality: DataQualitySummary
    last_updated: str

File: frontend/src/types/farmer.ts
export interface SoilValueWithProvenance {
  value?: any;
  provenance: 'MEASURED' | 'ESTIMATED' | 'UNKNOWN';
}

export interface GeoJSONPolygon {
  type: 'Polygon';
  coordinates: number[][][]; // [[[lng, lat], [lng, lat], ...]]
}

export interface PlantObservationRecord {
  id: number;
  field_id: number;
  crop_planting_id?: number;
  observation_date: string;
  image_url?: string;
  disease_result?: string;
  pest_result?: string;
  severity: string;
  notes?: string;
  location_in_field?: string;
  created_at?: string;
}

export interface CropPlanting {
  id: number;
  field_id: number;
  crop_id?: string;
  crop_name: string;
  scientific_name?: string;
  variety?: string;
  category?: string;
  sowing_date?: string;
  expected_harvest_date?: string;
  growth_stage?: string;
  growth_stage_source?: 'farmer' | 'calculated';
  cultivation_type?: string;
  irrigation_method?: string;
  water_availability?: string;
  status: 'PLANNED' | 'ACTIVE' | 'HARVESTED' | 'COMPLETED';
  notes?: string;
  created_at?: string;
  updated_at?: string;
}

export interface FieldRecord {
  id: number;
  farm_id: number;
  field_name: string;
  area_value: number;
  area_unit: string;
  total_area_m2: number;
  latitude?: number;
  longitude?: number;
  boundary_geojson?: GeoJSONPolygon | null;
  perimeter_m?: number;
  centroid_lat?: number;
  centroid_lng?: number;
  geometry_source?: 'GEOMETRIC' | 'MANUAL';
  geometry_updated_at?: string;
  soil_type?: string;
  soil_test_available: boolean;
  ph?: number;
  nitrogen?: number;
  phosphorus?: number;
  potassium?: number;
  organic_carbon?: number;
  water_source?: string;
  soil_data?: {
    ph: SoilValueWithProvenance;
    nitrogen: SoilValueWithProvenance;
    phosphorus: SoilValueWithProvenance;
    potassium: SoilValueWithProvenance;
    organic_carbon: SoilValueWithProvenance;
    ec: SoilValueWithProvenance;
    texture: SoilValueWithProvenance;
    moisture: SoilValueWithProvenance;
  };
  notes?: string;
  plantings: CropPlanting[];
  observations?: PlantObservationRecord[];
  created_at?: string;
  updated_at?: string;
}

export interface FarmRecord {
  id: number;
  farmer_id: number;
  farm_name: string;
  location_name?: string;
  latitude: number;
  longitude: number;
  area_value: number;
  area_unit: string;
  total_area_m2: number;
  soil_type_manual?: string;
  water_source?: string;
  irrigation_method?: string;
  ownership_type?: string;
  notes?: string;
  fields: FieldRecord[];
  created_at?: string;
  updated_at?: string;
}

export interface FarmerProfile {
  id: number;
  user_id: string;
  full_name: string;
  phone?: string;
  email?: string;
  preferred_language: string;
  farms: FarmRecord[];
  created_at?: string;
  updated_at?: string;
}

export interface RiskItem {
  id: string;
  category: string;
  title: string;
  description: string;
  severity: 'low' | 'medium' | 'high' | 'critical';
  priority: 'low' | 'medium' | 'high' | 'critical';
  status: string;
  confidence: string;
  reasoning: string;
  recommended_follow_up?: string;
  affected_crop?: string;
  affected_stage?: string;
  evidence?: any[];
  detected_at?: string;
  valid_until?: string;
}

export interface OpportunityItem {
  id: string;
  category: string;
  title: string;
  description: string;
  priority: 'low' | 'medium' | 'high' | 'critical';
  status: string;
  confidence: string;
  reasoning: string;
  suggested_action?: string;
  time_window?: string;
  affected_crop?: string;
  affected_stage?: string;
  evidence?: any[];
  detected_at?: string;
}

export interface ActionPlanItem {
  id: string;
  action_type: string;
  priority: 'low' | 'medium' | 'high' | 'critical';
  status: 'active' | 'monitoring' | 'needs_review' | 'TODO' | 'DONE' | 'DISMISSED';
  title: string;
  action: string;
  reason: string;
  affected_crop?: string;
  affected_stage?: string;
  location?: string;
  recommended_time?: string;
  time_window?: string;
  due_date?: string;
}

export interface SmartAlertItem {
  id: string;
  priority: 'Critical' | 'High' | 'Moderate' | 'Info';
  category: string;
  title: string;
  description: string;
  trigger_evidence?: string;
  potential_impact?: string;
  recommended_action?: string;
  status?: string;
  crop_name?: string;
  field_name?: string;
  timestamp: string;
  actionable: boolean;
}

export interface NotificationPreferences {
  farmer_id: number;
  channels: {
    in_app: boolean;
    email: boolean;
    sms: boolean;
    whatsapp: boolean;
  };
  categories: {
    critical_risks: boolean;
    weather: boolean;
    crop_health: boolean;
    irrigation: boolean;
    market: boolean;
    calendar: boolean;
    action_reminders: boolean;
  };
  quiet_hours: {
    enabled: boolean;
    start: string;
    end: string;
    critical_override: boolean;
  };
}

export interface WeatherImpactItem {
  crop_name: string;
  field_name: string;
  temperature?: number;
  rain_forecast_mm?: number;
  impact: string;
  action: string;
  status: 'Favorable' | 'Monitor' | 'Warning';
}

export interface SoilImpactItem {
  crop_name: string;
  field_name: string;
  ph_status: string;
  texture_status: string;
  n_status: string;
  p_status: string;
  k_status: string;
  impact_text: string;
}

export interface IrrigationContextItem {
  crop_name: string;
  field_name: string;
  current_moisture?: number;
  expected_moisture?: number;
  rainfall_forecast_mm?: number;
  irrigation_method?: string;
  water_availability?: string;
  status: 'LOW' | 'NORMAL' | 'HIGH' | 'MONITOR';
  next_window?: string;
  model_note: string;
}

export interface FertilizerContextItem {
  crop_name: string;
  field_name: string;
  n_status: string;
  p_status: string;
  k_status: string;
  model_available: boolean;
  model_scope_note: string;
  recommendation?: string;
  reason: string;
}

export interface PestRiskItem {
  crop_name: string;
  field_name: string;
  risk_level: 'Low' | 'Moderate' | 'High';
  weather_drivers: string[];
  crop_stage?: string;
  action: string;
  model_scope_note: string;
}

export interface MarketWatchItem {
  crop_name: string;
  commodity: string;
  current_price?: number;
  trend: string;
  change_30d_pct?: number;
  period: string;
  market_location?: string;
}

export interface CropTimelineItem {
  date_label: string;
  crop_name: string;
  field_name: string;
  event_title: string;
  reason: string;
  priority: 'Critical' | 'High' | 'Moderate' | 'Info';
  source: string;
  evidence_status: string;
}

export interface ImpactMatrixRow {
  crop_name: string;
  field_name: string;
  area_display: string;
  weather: string;
  soil: string;
  water: string;
  pest: string;
  market: string;
  attention_level: 'High' | 'Medium' | 'Low';
}

export interface DataQualitySummary {
  location_confidence: string;
  weather_freshness: string;
  soil_availability: string;
  npk_availability: string;
  market_status: string;
  profile_completeness_pct: number;
  missing_fields: string[];
}

export interface TodayFarmStatus {
  weather_summary: string;
  soil_summary: string;
  water_summary: string;
  market_summary: string;
  action_items_count: number;
  critical_risks_count?: number;
}

export interface ActiveCropCard {
  id: number;
  field_id: number;
  crop_name: string;
  field_name: string;
  area_display: string;
  growth_stage: string;
  days_since_sowing?: number;
  sowing_date?: string;
  expected_harvest?: string;
  weather_status: string;
  water_status: string;
  pest_status: string;
  market_trend: string;
}

export interface FarmDashboardResponse {
  farmer: FarmerProfile;
  location: {
    latitude: number;
    longitude: number;
    display_name: string;
    source: string;
  };
  total_farm_area: number;
  total_farm_area_unit: string;
  active_crops_count: number;
  fields_count: number;
  today_status: TodayFarmStatus;
  active_crop_cards: ActiveCropCard[];
  weather_impacts: WeatherImpactItem[];
  soil_impacts: SoilImpactItem[];
  irrigation_items: IrrigationContextItem[];
  fertilizer_items: FertilizerContextItem[];
  pest_items: PestRiskItem[];
  market_watch: MarketWatchItem[];
  crop_calendar_events: any[];
  risks_and_opportunities?: {
    risks: RiskItem[];
    opportunities: OpportunityItem[];
    summary?: any;
  };
  risks?: RiskItem[];
  opportunities?: OpportunityItem[];
  actions?: ActionPlanItem[];
  risk_opportunity?: any;
  action_plan?: {
    actions: ActionPlanItem[];
    total_actions: number;
  };
  alerts: SmartAlertItem[];
  dispatched_alerts?: SmartAlertItem[];
  plant_observations?: PlantObservationRecord[];
  notification_preferences?: NotificationPreferences;
  timeline: CropTimelineItem[];
  impact_matrix: ImpactMatrixRow[];
  data_quality: DataQualitySummary;
  last_updated: string;
}
